tests/sentry/api/endpoints/test_project_details.py

```diff
@@ -783,1120 +777,1068 @@ def test_invalid_digests_max_delay(self):
-        # change first rule
-        saved_config["rules"][0]["sampleRate"] = 0.1
-        # do not touch the second rule
-        # remove third rule (the id should NEVER be reused)
-        del saved_config["rules"][2]
-        # insert a new element at position 0
-        new_rule_1 = {
-            "sampleRate": 0.22,
-            "type": "trace",
-            "condition": {
-                "op": "and",
-                "inner": [
-                    {"op": "eq", "name": "field1", "value": ["val"]},
-                ],
-            },
-            "id": -1,
+            ],
+            "next_id": 3,
         }
 
-        saved_config["rules"].insert(0, new_rule_1)
-        # insert a new element at the end
-        new_rule_2 = {
-            "sampleRate": 0.33,
-            "type": "trace",
-            "condition": {
-                "op": "and",
-                "inner": [],
-            },
-            "id": -1,
-        }
+        serializer = DynamicSamplingSerializer(
+            data=data,
+            context={"project": self.create_project(), "request": self.make_request()},
+        )
+        assert serializer.is_valid()
+        assert data == serializer.validated_data
 
-        saved_config["rules"].append(new_rule_2)
-        with Feature({"organizations:server-side-sampling": True}):
-            # turn it back from ordered dict to dict (both main obj and rules)
-            saved_config = dict(saved_config)
-            saved_config["rules"] = [dict(rule) for rule in saved_config["rules"]]
-            self.get_success_response(self.org_slug, self.proj_slug, dynamicSampling=saved_config)
-        response = self.get_success_response(self.org_slug, self.proj_slug, method="get")
-        saved_config = response.data["dynamicSampling"]
-        new_ids = [rule["id"] for rule in saved_config["rules"]]
-        # first rule is new, second rule got a new id because it is changed,
-        # third rule (used to be second) keeps the id, fourth rule is new
-        assert new_ids == [4, 5, 2, 6]
-        new_next_id = saved_config["next_id"]
-        assert new_next_id == 7
 
-    def test_dynamic_sampling_rules_have_active_flag(self):
-        """
-        Tests that the active flag is set for all rules
-        """
-        data = _dyn_sampling_data()
-        with Feature({"organizations:server-side-sampling": True}):
-            self.get_success_response(self.org_slug, self.proj_slug, dynamicSampling=data)
-            response = self.get_success_response(self.org_slug, self.proj_slug, method="get")
-        saved_config = response.data["dynamicSampling"]
-        assert all([rule["active"] for rule in saved_config["rules"]])
-        assert AuditLogEntry.objects.filter(
-            organization=self.project.organization,
-            target_object=self.project.id,
-            event=audit_log.get_event_id("SAMPLING_RULE_ADD"),
-        ).exists()
+class TestProjectDetailsDynamicSamplingBase(APITestCase, ABC):
```

Comment:
why do we need ABC here? it's not clear to me. 🤔

Reply:
simply because it s a base class

Reply:
oh i see i mis read diff. Cool resolved